python/ray/data/datasource/text_datasource.py

```diff
@@ -21,10 +22,18 @@ def _read_file(
         assert len(block) == 1
         data = block[0]
 
+        builder = DelegatingBlockBuilder()
+
+        drop_empty_lines = reader_args["drop_empty_lines"]
         lines = data.decode(reader_args["encoding"]).split("\n")
-        if reader_args["drop_empty_lines"]:
-            lines = [line for line in lines if line.strip() != ""]
-        return lines
+        for line in lines:
+            if drop_empty_lines and line.strip() == "":
+                continue
+            item = {self._COLUMN_NAME: line}
+            builder.add(item)
+
+        block = builder.build()
+        return block
 
     def _convert_block_to_tabular_block(
```

Comment:
actually this doesn't work as is.
for some reasons, TextDatasource inherits BinaryDatasource not FileBasedDatasource.
BinaryDatasource requires block size to be 1 when you convert it to tabular data.
If I switch it to inherit from FileBasedDatasource, it seems like we need to implement a lot more APIs, read/write, etc.

Reply:
got it. This `BinaryDatasource` becomes not that useful. We can keep `_convert_block_to_tabular_block` here for now.

Reply:
This inherits from `BinaryDatasource` in order to share the `_read_file()` implementation, since the text datasource is currently a thin decoding layer on top of reading binary data. If that proves to be cumbersome, we could look at a different code-sharing method between the binary and text datasources.

Reply:
@clarkzinzow - I am thinking about to move the `_read_file()` into a utility method instead. Then `ImageDatasource` and `TextDatasource` does not need to inherit from `BinaryDatasource`, and we can delete `_convert_block_to_tabular_block ` from `ImageDatasource` and `TextDatasource`. I can take a stab to do it.

Reply:
Yep same thought! And sounds good.

Reply:
@gjoliver - feel free to merge for this PR as it is. I will take refactoring part in another PR.